server/src/main/java/org/tctalent/server/service/db/impl/CandidateServiceImpl.java

```diff
@@ -405,2392 +405,2410 @@
+    }
+  }
+
+  @Override
+  public Optional<Candidate> getLoggedInCandidateLoadCertifications() {
+    Long candidateId = authService.getLoggedInCandidateId();
+    if (candidateId == null) {
+      return Optional.empty();
+    } else {
+      Candidate candidate = candidateRepository
+          .findByIdLoadCertifications(candidateId);
+      return candidate == null ? Optional.empty() : Optional.of(candidate);
+    }
+  }
+
+  @Override
+  public Optional<Candidate> getLoggedInCandidateLoadCandidateLanguages() {
+    Long candidateId = authService.getLoggedInCandidateId();
+    if (candidateId == null) {
+      return Optional.empty();
+    } else {
+      Candidate candidate = candidateRepository
+          .findByIdLoadCandidateLanguages(candidateId);
+      return candidate == null ? Optional.empty() : Optional.of(candidate);
+    }
+  }
+
+  @Override
+  public Optional<Candidate> getLoggedInCandidate() {
+    User user = authService.getLoggedInUser().orElse(null);
+    if (user == null) {
+      return Optional.empty();
+    }
+    Candidate candidate = candidateRepository.findByUserId(user.getId());
+
+    populateTransientTaskAssignmentFields(candidate.getTaskAssignments());
+
+    return Optional.of(candidate);
+  }
+
+  @Override
+  public Candidate findByCandidateNumber(String candidateNumber) {
+    return candidateRepository.findByCandidateNumber(candidateNumber);
+  }
+
+  @Override
+  public List<Candidate> findByIds(@NonNull Collection<Long> ids) {
+    return candidateRepository.findByIds(ids);
+  }
+
+  @Override
+  public Candidate findByCandidateNumberRestricted(String candidateNumber) {
+    User loggedInUser = authService.getLoggedInUser()
+        .orElseThrow(() -> new InvalidSessionException("Not logged in"));
+
+    Set<Country> sourceCountries = userService.getDefaultSourceCountries(loggedInUser);
+    return candidateRepository.findByCandidateNumberRestricted(candidateNumber, sourceCountries)
+        .orElseThrow(() -> new CountryRestrictionException("You don't have access to this candidate."));
+  }
+
+  @Override
+  public Candidate findByIdLoadSavedLists(long candidateId) {
+    return candidateRepository.findByIdLoadSavedLists(candidateId);
+  }
+
+  @Override
+  public Candidate findByIdLoadUser(long id, Set<Country> sourceCountries) {
+    return candidateRepository.findByIdLoadUser(id, sourceCountries).orElse(null);
+  }
+
+  @Transactional(readOnly = true)
+  public void validateContactRequest(User user, BaseCandidateContactRequest request) {
+    // Check email not already taken
+    if (!StringUtils.isBlank(request.getEmail())) {
+      try {
+        User exists = userRepository.findByEmailIgnoreCase(request.getEmail());
+        if (user == null && exists != null || exists != null && !exists.getId().equals(user.getId())) {
+          throw new UsernameTakenException("email");
+        }
+      } catch (IncorrectResultSizeDataAccessException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)